GVRf/Framework/jni/engine/importer/assimp_importer.cpp

```diff
@@ -68,4 +71,176 @@ std::shared_ptr<Mesh> AssimpImporter::getMesh(int index) {
 
     return std::shared_ptr < Mesh > (mesh);
 }
+
+void AssimpImporter::sceneRecursion(aiNode* node, const aiScene* aScene, std::shared_ptr<Scene> scnPtr, JNIEnv * env, jobject obj, jobject bitmap, aiMatrix4x4 accumulatedTransform)
+{
+    for(int i=0; i < node->mNumMeshes; i++)
+    {
+        std::shared_ptr<SceneObject> scnObject(new SceneObject());
+
+        // Creates a scene object for the node children meshes.
+        if(scnObject->render_data() == NULL)
+        {
+            // Scene Object has no render data.
+
+            // Mesh
+            std::shared_ptr<Mesh> mesh = getMesh(node->mMeshes[i]);
+
+            // New render data object.
+            std::shared_ptr<RenderData> renderData(new RenderData());
+
+            // Set the mesh to the render data.
+            renderData->set_mesh(mesh);
+
+            // Set material.
+            aiMesh* ai_mesh = aScene->mMeshes[node->mMeshes[i]];
+            aiMaterial* ai_material = aScene->mMaterials[ai_mesh->mMaterialIndex];
+
+            // Defines a shader with shader type 0. UNLIT_SHADER = 0
+            std::shared_ptr<Material> material(new Material(static_cast<Material::ShaderType>(0)));
+
+            // Actual texture image
+            aiString textureImage;
+            ai_material->GetTexture(aiTextureType_DIFFUSE, i, &textureImage);
+            jstring texFileName = env->NewStringUTF(textureImage.C_Str());
+            jclass texClass = env->FindClass("org/gearvrf/GVRContext");
+
+            if (texClass == NULL) {
+                // Couldn't find the class GVRContext.
+            } else {
+                jmethodID mID = env->GetStaticMethodID(texClass, "loadBitmap", "(Ljava/lang/String;)Landroid/graphics/Bitmap;");
+                if (mID == NULL) {
+                    // No such method in Java side
+                    // No interaction with Java side so apply default texture
+                    std::shared_ptr<Texture> texture(new BaseTexture(env, bitmap));
+                    material->setTexture("main_texture", texture);
+                    // Sets the default main texture
+
+                } else {
+                    if ((ai_material->GetTextureCount(aiTextureType_DIFFUSE)) <= 0) {
+                        // No texture file found so apply default texture
+                        std::shared_ptr<Texture> texture(new BaseTexture(env, bitmap));
+                        material->setTexture("main_texture", texture);
+                    } else {
+                        // About to enter Java side.
+                        jobject mBitmap = env->CallStaticObjectMethod(texClass, mID, texFileName);
```

Comment:
1. This code is assuming that `texFileName` contains the name of a file in the `assets` directory. Is this always true? What if the mesh file is in a subdirectory of the `assets` directory? What if the  mesh file is in `res/raw`?

2. `GVRContext.loadBitmap(String)` can crash the app, if the bitmap is too large, or if memory is fragmented. `GVRContext.loadTexture(GVRAndroidResource)` is safer and will often be faster.